### ISSUE TITLE
Fake News Detection System

### ISSUE THREAD

**Issue by sameed6152** [aasemquazi/test1#5]
# ============================================
# Fake News Detection System using AI
# ============================================

# Install Required Libraries (Run Once)
# Uncomment if libraries are not installed

# !pip install pandas
# !pip install numpy
# !pip install scikit-learn
# !pip install nltk

# ============================================
# Import Libraries
# ============================================

import pandas as pd
import numpy as np
import string
import nltk

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# Download stopwords
nltk.download('stopwords')

from nltk.corpus import stopwords

# ============================================
# Load Dataset
# ============================================

# Dataset file name
# Replace with your dataset file path

df = pd.read_csv("news.csv")

# Show first 5 rows
print(df.head())

# ============================================
# Check Dataset Information
# ============================================

print("\nDataset Shape:")
print(df.shape)

print("\nColumns:")
print(df.columns)

# ============================================
# Data Preprocessing Function
# ============================================

stop_words = set(stopwords.words('english'))

def clean_text(text):

    # Convert to lowercase
    text = text.lower()

    # Remove punctuation
    text = "".join([char for char in text if char not in string.punctuation])

    # Remove stopwords
    words = text.split()

    words = [word for word in words if word not in stop_words]

    # Join words again
    text = " ".join(words)

    return text

# Apply cleaning function
df['text'] = df['text'].apply(clean_text)

# ============================================
# Features and Labels
# ============================================

X = df['text']
y = df['label']

# ============================================
# Convert Text into Numerical Data
# ============================================

tfidf_vectorizer = TfidfVectorizer(max_df=0.7)

tfidf_train = tfidf_vectorizer.fit_transform(X)

# ============================================
# Split Dataset
# ============================================

X_train, X_test, y_train, y_test = train_test_split(
    tfidf_train,
    y,
    test_size=0.2,
    random_state=42
)

# ============================================
# Train Machine Learning Model
# ============================================

model = PassiveAggressiveClassifier(max_iter=50)

model.fit(X_train, y_train)

# ============================================
# Make Predictions
# ============================================

y_pred = model.predict(X_test)

# ============================================
# Calculate Accuracy
# ============================================

score = accuracy_score(y_test, y_pred)

print("\nAccuracy:", round(score * 100, 2), "%")

# ============================================
# Confusion Matrix
# ============================================

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# ============================================
# Test Custom News
# ============================================

news = input("\nEnter News Text: ")

# Clean input text
clean_news = clean_text(news)

# Convert into vector
vector_input = tfidf_vectorizer.transform([clean_news])

# Predict
prediction = model.predict(vector_input)

# Show Result
if prediction[0] == 'FAKE':
    print("\nThis News is FAKE")
else:
    print("\nThis News is REAL")

# ============================================
# End of Project
# ============================================
